admin: add command to update go packages with norm_version

crozzy opened the pull request:
There are a number of Go packages in the `package` table that do not have the needed norm_version which is required for Go matching.

This looks up the potential Go packages from the `matcher.vuln` table, the difficultly doing this is entirely in the indexer DB is how to identify that these packages came from the gobin scanner. Trying to use the `package_scanartifact` (joining to the `scanners` table is prohibitively expensive as the query path is not accounted for in the index (i.e. we're always expecting the first clause to the the layer_id).

crozzy committed:
admin: add command to update go packages with norm_version

There are a number of Go packages in the `package` table that
do not have the needed norm_version which is required for Go
matching.

Signed-off-by: crozzy <joseph.crosland@gmail.com>

crozzy commented:
Local output:
```
 ► go build -o . ./cmd/clairctl && ./clairctl -D -c local-dev/clair/config.yaml admin oneoff update-golang-packages
2023-11-16T10:00:32-08:00 INF using discovered connection params database=matcher host=localhost port=32775 user=clair
2023-11-16T10:00:32-08:00 DBG resizing pool pool size=2
2023-11-16T10:00:32-08:00 INF using discovered connection params database=indexer host=localhost port=32775 user=clair
2023-11-16T10:00:32-08:00 DBG resizing pool pool size=2
2023-11-16T10:00:32-08:00 DBG working on version package_name=k8s.io/apiserver version=v0.27.4
2023-11-16T10:00:32-08:00 INF successfully updated package row package_id=3818 package_name=k8s.io/apiserver rows affected=1 version=v0.27.4
2023-11-16T10:00:32-08:00 DBG working on version package_name=stdlib version="go1.20.10 X:strictfipsruntime"
2023-11-16T10:00:32-08:00 WRN error parsing semver error="Invalid Semantic Version" package_name=stdlib version="go1.20.10 X:strictfipsruntime"
```

hdonnay commented:
Do we want this to be `go` specific, or couch it as a general "fix-up vulnerability table" command?
I don't feel strongly, other than my initial instinct was to do the latter.

crozzy commented:
> Do we want this to be `go` specific, or couch it as a general "fix-up vulnerability table" command? I don't feel strongly, other than my initial instinct was to do the latter.

I think the issue with it not being go-specific is; is the version parsing logic and the `norm_kind` value applicable to any other ecosystems? Most matchers aren't relying on that value, are you thinking that we could make it general purpose and convert some more matching logic over to do `func (matcher *Matcher) VersionAuthoritative() bool { return true }`?

hdonnay commented:
I meant for the command itself -- are we ever going to need another `update-someother-package` command?

crozzy committed:
admin: add a check for compatible migration version

Add the ability to check migration version of the database before
running an admin command so you know you are operating on a compatible
database schema.

Signed-off-by: crozzy <joseph.crosland@gmail.com>

hdonnay reviewed (dismissed):
LGTM
